GTK 3.0 development libraries are required for Linux development. [Ubuntu 18.04]

## Flutter Doctor Output

```
tirth@piedcipher:~$ flutter doctor
Doctor summary (to see all details, run flutter doctor -v):
[✓] Flutter (Channel dev, 1.21.0-9.0.pre, on Linux, locale en_IN)
 
[✓] Android toolchain - develop for Android devices (Android SDK version 29.0.2)
[✓] Chrome - develop for the web
[✗] Linux toolchain - develop for Linux desktop
    ✗ GTK 3.0 development libraries are required for Linux development.
      They are likely available from your distribution (e.g.: apt install libgtk-3-dev)
[✓] Android Studio (version 4.0)
[✓] IntelliJ IDEA Ultimate Edition (version 2019.3)
[✓] VS Code (version 1.47.3)
[✓] Connected device (3 available)

! Doctor found issues in 1 category.
tirth@piedcipher:~$ gedit ~/.bashrc 
tirth@piedcipher:~$ clear

tirth@piedcipher:~$ flutter doctor -v
[✓] Flutter (Channel dev, 1.21.0-9.0.pre, on Linux, locale en_IN)
    • Flutter version 1.21.0-9.0.pre at /home/tirth/flutter
    • Framework revision 7c6f9dd239 (5 days ago), 2020-08-09 10:31:03 -0400
    • Engine revision 6d86e67f04
    • Dart version 2.10.0 (build 2.10.0-4.0.dev 0341576448)

 
[✓] Android toolchain - develop for Android devices (Android SDK version 29.0.2)
    • Android SDK at /home/tirth/Android/Sdk
    • Platform android-29, build-tools 29.0.2
    • Java binary at: /home/tirth/android-studio/jre/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6222593)
    • All Android licenses accepted.

[✓] Chrome - develop for the web
    • Chrome at google-chrome

[✗] Linux toolchain - develop for Linux desktop
    • clang version 6.0.0-1ubuntu2 (tags/RELEASE_600/final)
    • cmake version 3.10.2
    • ninja version 1.8.2
    • pkg-config version 0.29.1
    ✗ GTK 3.0 development libraries are required for Linux development.
      They are likely available from your distribution (e.g.: apt install libgtk-3-dev)

[✓] Android Studio (version 4.0)
    • Android Studio at /home/tirth/android-studio
    • Flutter plugin version 48.1.2
    • Dart plugin version 193.7361
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6222593)

[✓] IntelliJ IDEA Ultimate Edition (version 2019.3)
    • IntelliJ at /home/tirth/idea-IU-193.6015.39
    • Flutter plugin version 45.1.2
    • Dart plugin version 193.6015.53

[✓] VS Code (version 1.47.3)
    • VS Code at /usr/share/code
    • Flutter extension version 3.13.2

[✓] Connected device (3 available)
    • Linux (desktop)  • linux      • linux-x64      • Linux
    • Web Server (web) • web-server • web-javascript • Flutter Tools
    • Chrome (web)     • chrome     • web-javascript • Google Chrome 84.0.4147.125

! Doctor found issues in 1 category.
```

## Output for sudo apt install libgtk-3-dev

```
tirth@piedcipher:~$ sudo apt install libgtk-3-dev
Reading package lists... Done
Building dependency tree       
Reading state information... Done
libgtk-3-dev is already the newest version (3.22.30-1ubuntu4).
The following packages were automatically installed and are no longer required:
  efibootmgr libappstream-glib8 libavahi-ui-gtk3-0 libfreerdp-client2-2 libfreerdp2-2 libfwup1 libllvm8 libllvm9 libllvm9:i386 libnvidia-common-390
  libphodav-2.0-0 libphodav-2.0-common libspice-client-glib-2.0-8 libspice-client-gtk-3.0-5 libssh-4 libusbredirhost1 libvncclient1 libwinpr2-2
  spice-client-glib-usb-acl-helper
Use 'sudo apt autoremove' to remove them.
0 upgraded, 0 newly installed, 0 to remove and 0 not upgraded.
```

## Output for cat /etc/os-release

```
tirth@piedcipher:~$ cat /etc/os-release
NAME="Ubuntu"
VERSION="18.04.5 LTS (Bionic Beaver)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 18.04.5 LTS"
VERSION_ID="18.04"
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
VERSION_CODENAME=bionic
UBUNTU_CODENAME=bionic
```

Can anyone please help me?

@stuartmorgan I tried `sudo apt install libblkid-dev`. After that `flutter doctor` issue is gone.

```
tirth@piedcipher:~$ flutter doctor -v
[✓] Flutter (Channel dev, 1.21.0-9.0.pre, on Linux, locale en_IN)
    • Flutter version 1.21.0-9.0.pre at /home/tirth/flutter
    • Framework revision 7c6f9dd239 (5 days ago), 2020-08-09 10:31:03 -0400
    • Engine revision 6d86e67f04
    • Dart version 2.10.0 (build 2.10.0-4.0.dev 0341576448)

 
[✓] Android toolchain - develop for Android devices (Android SDK version 29.0.2)
    • Android SDK at /home/tirth/Android/Sdk
    • Platform android-29, build-tools 29.0.2
    • Java binary at: /home/tirth/android-studio/jre/bin/java
    • Java version OpenJDK Runtime Environment (build
      1.8.0_242-release-1644-b3-6222593)
    • All Android licenses accepted.

[✓] Chrome - develop for the web
    • Chrome at google-chrome

[✓] Linux toolchain - develop for Linux desktop
    • clang version 6.0.0-1ubuntu2 (tags/RELEASE_600/final)
    • cmake version 3.10.2
    • ninja version 1.8.2
    • pkg-config version 0.29.1

[✓] Android Studio (version 4.0)
    • Android Studio at /home/tirth/android-studio
    • Flutter plugin version 48.1.2
    • Dart plugin version 193.7361
    • Java version OpenJDK Runtime Environment (build
      1.8.0_242-release-1644-b3-6222593)

[✓] IntelliJ IDEA Ultimate Edition (version 2019.3)
    • IntelliJ at /home/tirth/idea-IU-193.6015.39
    • Flutter plugin version 45.1.2
    • Dart plugin version 193.6015.53

[✓] VS Code (version 1.47.3)
    • VS Code at /usr/share/code
    • Flutter extension version 3.13.2

[✓] Connected device (3 available)
    • Linux (desktop)  • linux      • linux-x64      • Linux
    • Web Server (web) • web-server • web-javascript • Flutter Tools
    • Chrome (web)     • chrome     • web-javascript • Google Chrome
      84.0.4147.125

• No issues found!
```

But I'm getting a CMake Error while building the app.

![Screenshot at 2020-08-14 21-40-51](https://user-images.githubusercontent.com/13456345/90269852-ee13be80-de76-11ea-8a21-fa55eb6fc88d.png)

--- a/packages/flutter_tools/lib/src/base/user_messages.dart
+++ b/packages/flutter_tools/lib/src/base/user_messages.dart
@@ -237,6 +237,8 @@ class UserMessages {
   String pkgConfigTooOld(String minimumVersion) => 'pkg-config $minimumVersion or later is required.';
   String get gtkLibrariesMissing => 'GTK 3.0 development libraries are required for Linux development.\n'
       'They are likely available from your distribution (e.g.: apt install libgtk-3-dev)';
+  String get blkidLibraryMissing => 'The blkid development library is required for Linux development.\n'
+      'It is likely available from your distribution (e.g.: apt install libblkid-dev)';
 
   // Messages used in FlutterCommand
   String flutterElapsedTime(String name, String elapsedTime) => '"flutter $name" took $elapsedTime.';

--- a/packages/flutter_tools/lib/src/linux/linux_doctor.dart
+++ b/packages/flutter_tools/lib/src/linux/linux_doctor.dart
@@ -52,11 +52,10 @@ class LinuxDoctorValidator extends DoctorValidator {
     kPkgConfigBinary: Version(0, 29, 0),
   };
 
-  final List<String> _requiredLibraries = <String>[
+  final List<String> _requiredGtkLibraries = <String>[
     'gtk+-3.0',
     'glib-2.0',
     'gio-2.0',
-    'blkid',
   ];
 
   @override
@@ -136,10 +135,10 @@ class LinuxDoctorValidator extends DoctorValidator {
       }
     }
 
-    // Message for libraries.
+    // Messages for libraries.
     {
       bool libraryMissing = false;
-      for (final String library in _requiredLibraries) {
+      for (final String library in _requiredGtkLibraries) {
         if (!await _libraryIsPresent(library)) {
           libraryMissing = true;
           break;
@@ -149,6 +148,10 @@ class LinuxDoctorValidator extends DoctorValidator {
         validationType = ValidationType.missing;
         messages.add(ValidationMessage.error(_userMessages.gtkLibrariesMissing));
       }
+    }
+    if (!await _libraryIsPresent('blkid')) {
+      validationType = ValidationType.missing;
+      messages.add(ValidationMessage.error(_userMessages.blkidLibraryMissing));
     }
 
     return ValidationResult(validationType, messages);

--- a/packages/flutter_tools/test/general.shard/linux/linux_doctor_test.dart
+++ b/packages/flutter_tools/test/general.shard/linux/linux_doctor_test.dart
@@ -55,26 +55,31 @@ FakeCommand _pkgConfigPresentCommand(String version) {
   );
 }
 
-// Commands that give positive replies for all the library pkg-config queries.
-List<FakeCommand> _librariesPresentCommands() {
-  return const <FakeCommand>[
-    FakeCommand(command: <String>['pkg-config', '--exists', 'gtk+-3.0']),
-    FakeCommand(command: <String>['pkg-config', '--exists', 'glib-2.0']),
-    FakeCommand(command: <String>['pkg-config', '--exists', 'gio-2.0']),
-    FakeCommand(command: <String>['pkg-config', '--exists', 'blkid']),
+/// A command that returns either success or failure for a pkg-config query
+/// for [library], depending on [exists].
+FakeCommand _libraryCheckCommand(String library, {bool exists = true}) {
+  return FakeCommand(
+    command: <String>['pkg-config', '--exists', library],
+    exitCode: exists ? 0 : 1,
+  );
+}
+
+// Commands that give positive replies for all the GTK library pkg-config queries.
+List<FakeCommand> _gtkLibrariesPresentCommands() {
+  return <FakeCommand>[
+    _libraryCheckCommand('gtk+-3.0'),
+    _libraryCheckCommand('glib-2.0'),
+    _libraryCheckCommand('gio-2.0'),
   ];
 }
 
-// Commands that give some failures for the library pkg-config queries.
-List<FakeCommand> _librariesMissingCommands() {
-  return const <FakeCommand>[
-    FakeCommand(command: <String>['pkg-config', '--exists', 'gtk+-3.0']),
-    FakeCommand(
-      command: <String>['pkg-config', '--exists', 'glib-2.0'],
-      exitCode: 1,
-    ),
-    FakeCommand(command: <String>['pkg-config', '--exists', 'gio-2.0']),
-    FakeCommand(command: <String>['pkg-config', '--exists', 'blkid']),
+// Commands that give some failures for the GTK library pkg-config queries.
+List<FakeCommand> _gtkLibrariesMissingCommands() {
+  return <FakeCommand>[
+    _libraryCheckCommand('gtk+-3.0'),
+    _libraryCheckCommand('glib-2.0', exists: false),
+    // No more entries, since the first missing GTK library stops the
+    // checks.
   ];
 }
 
@@ -93,7 +98,8 @@ void main() {
       _cmakePresentCommand('3.16.3'),
       _ninjaPresentCommand('1.10.0'),
       _pkgConfigPresentCommand('0.29'),
-      ..._librariesPresentCommands(),
+      ..._gtkLibrariesPresentCommands(),
+      _libraryCheckCommand('blkid'),
     ]);
     final DoctorValidator linuxDoctorValidator = LinuxDoctorValidator(
       processManager: processManager,
@@ -116,7 +122,8 @@ void main() {
       _cmakePresentCommand('3.16.3'),
       _ninjaPresentCommand('1.10.0'),
       _pkgConfigPresentCommand('0.29'),
-      ..._librariesPresentCommands(),
+      ..._gtkLibrariesPresentCommands(),
+      _libraryCheckCommand('blkid'),
     ]);
     final DoctorValidator linuxDoctorValidator = LinuxDoctorValidator(
       processManager: processManager,
@@ -140,7 +147,8 @@ void main() {
       _cmakePresentCommand('3.2.0'),
       _ninjaPresentCommand('1.10.0'),
       _pkgConfigPresentCommand('0.29'),
-      ..._librariesPresentCommands(),
+      ..._gtkLibrariesPresentCommands(),
+      _libraryCheckCommand('blkid'),
     ]);
     final DoctorValidator linuxDoctorValidator = LinuxDoctorValidator(
       processManager: processManager,
@@ -164,7 +172,8 @@ void main() {
       _cmakePresentCommand('3.16.3'),
       _ninjaPresentCommand('0.8.1'),
       _pkgConfigPresentCommand('0.29'),
-      ..._librariesPresentCommands(),
+      ..._gtkLibrariesPresentCommands(),
+      _libraryCheckCommand('blkid'),
     ]);
     final DoctorValidator linuxDoctorValidator = LinuxDoctorValidator(
       processManager: processManager,
@@ -188,7 +197,8 @@ void main() {
       _cmakePresentCommand('3.16.3'),
       _ninjaPresentCommand('1.10.0'),
       _pkgConfigPresentCommand('0.27.0'),
-      ..._librariesPresentCommands(),
+      ..._gtkLibrariesPresentCommands(),
+      _libraryCheckCommand('blkid'),
     ]);
     final DoctorValidator linuxDoctorValidator = LinuxDoctorValidator(
       processManager: processManager,
@@ -212,7 +222,8 @@ void main() {
       _missingBinaryCommand('cmake'),
       _ninjaPresentCommand('1.10.0'),
       _pkgConfigPresentCommand('0.29'),
-      ..._librariesPresentCommands(),
+      ..._gtkLibrariesPresentCommands(),
+      _libraryCheckCommand('blkid'),
     ]);
     final UserMessages userMessages = UserMessages();
     final DoctorValidator linuxDoctorValidator = LinuxDoctorValidator(
@@ -236,7 +247,8 @@ void main() {
       _cmakePresentCommand('3.16.3'),
       _ninjaPresentCommand('1.10.0'),
       _pkgConfigPresentCommand('0.29'),
-      ..._librariesPresentCommands(),
+      ..._gtkLibrariesPresentCommands(),
+      _libraryCheckCommand('blkid'),
     ]);
     final UserMessages userMessages = UserMessages();
     final DoctorValidator linuxDoctorValidator = LinuxDoctorValidator(
@@ -260,7 +272,8 @@ void main() {
       _cmakePresentCommand('3.16.3'),
       _missingBinaryCommand('ninja'),
       _pkgConfigPresentCommand('0.29'),
-      ..._librariesPresentCommands(),
+      ..._gtkLibrariesPresentCommands(),
+      _libraryCheckCommand('blkid'),
     ]);
     final UserMessages userMessages = UserMessages();
     final DoctorValidator linuxDoctorValidator = LinuxDoctorValidator(
@@ -284,7 +297,8 @@ void main() {
       _cmakePresentCommand('3.16.3'),
       _ninjaPresentCommand('1.10.0'),
       _missingBinaryCommand('pkg-config'),
-      ..._librariesPresentCommands(),
+      ..._gtkLibrariesPresentCommands(),
+      _libraryCheckCommand('blkid'),
     ]);
     final UserMessages userMessages = UserMessages();
     final DoctorValidator linuxDoctorValidator = LinuxDoctorValidator(
@@ -302,13 +316,14 @@ void main() {
     ]);
   });
 
-  testWithoutContext('Missing validation when libraries are not available', () async {
+  testWithoutContext('Missing validation when GTK libraries are not available', () async {
     final ProcessManager processManager = FakeProcessManager.list(<FakeCommand>[
       _clangPresentCommand('4.0.1'),
       _cmakePresentCommand('3.16.3'),
       _ninjaPresentCommand('1.10.0'),
       _pkgConfigPresentCommand('0.29'),
-      ..._librariesMissingCommands(),
+      ..._gtkLibrariesMissingCommands(),
+      _libraryCheckCommand('blkid'),
     ]);
     final UserMessages userMessages = UserMessages();
     final DoctorValidator linuxDoctorValidator = LinuxDoctorValidator(
@@ -327,13 +342,40 @@ void main() {
     ]);
   });
 
+  testWithoutContext('Missing validation when libraries are not available', () async {
+    final ProcessManager processManager = FakeProcessManager.list(<FakeCommand>[
+      _clangPresentCommand('4.0.1'),
+      _cmakePresentCommand('3.16.3'),
+      _ninjaPresentCommand('1.10.0'),
+      _pkgConfigPresentCommand('0.29'),
+      ..._gtkLibrariesPresentCommands(),
+      _libraryCheckCommand('blkid', exists: false),
+    ]);
+    final UserMessages userMessages = UserMessages();
+    final DoctorValidator linuxDoctorValidator = LinuxDoctorValidator(
+      processManager: processManager,
+      userMessages: userMessages,
+    );
+    final ValidationResult result = await linuxDoctorValidator.validate();
+
+    expect(result.type, ValidationType.missing);
+    expect(result.messages, <ValidationMessage>[
+      const ValidationMessage('clang version 4.0.1-6+build1'),
+      const ValidationMessage('cmake version 3.16.3'),
+      const ValidationMessage('ninja version 1.10.0'),
+      const ValidationMessage('pkg-config version 0.29'),
+      ValidationMessage.error(userMessages.blkidLibraryMissing),
+    ]);
+  });
+
   testWithoutContext('Missing validation when multiple dependencies are not available', () async {
     final ProcessManager processManager = FakeProcessManager.list(<FakeCommand>[
       _missingBinaryCommand('clang++'),
       _missingBinaryCommand('cmake'),
       _ninjaPresentCommand('1.10.0'),
       _pkgConfigPresentCommand('0.29'),
-      ..._librariesPresentCommands(),
+      ..._gtkLibrariesPresentCommands(),
+      _libraryCheckCommand('blkid'),
     ]);
     final DoctorValidator linuxDoctorValidator = LinuxDoctorValidator(
       processManager: processManager,

